app/controllers/supervisor/base_controller.rb

```diff
@@ -0,0 +1,18 @@
+class Supervisor::BaseController < ApplicationController
+
+  before_action :must_be_supervisor
```

Comment:
I just realised/remembered this project uses the [cancancan](https://github.com/CanCanCommunity/cancancan) gem for exactly this kind of purpose. Could you possibly try and use that to authorize supervisors? If you get stuck, feel free to ping me.